python_modules/dagster/dagster_tests/asset_defs_tests/test_asset_deps.py

```diff
@@ -0,0 +1,399 @@
+    )
+    def downstream():
+        return None, None
+
+    assert len(downstream.input_names) == 3
+    assert downstream.op.ins["via_definition"].dagster_type.kind == DagsterTypeKind.NOTHING
+    assert downstream.op.ins["via_string"].dagster_type.kind == DagsterTypeKind.NOTHING
+    assert downstream.op.ins["via_asset_key"].dagster_type.kind == DagsterTypeKind.NOTHING
+
+    res = materialize(
+        [via_definition, via_string, via_asset_key, downstream],
+        resources={"io_manager": TestingIOManager()},
+    )
+    assert res.success
+
+
+def test_source_asset_deps_via_assets_definition():
+    a_source_asset = SourceAsset("a_key")
+
+    @asset(deps=[a_source_asset])
+    def depends_on_source_asset():
+        return None
+
+    assert len(depends_on_source_asset.input_names) == 1
+    assert depends_on_source_asset.op.ins["a_key"].dagster_type.kind == DagsterTypeKind.NOTHING
+
+    res = materialize([depends_on_source_asset], resources={"io_manager": TestingIOManager()})
+    assert res.success
+
+
+def test_source_asset_deps_via_string():
+    a_source_asset = SourceAsset("a_key")  # noqa: F841
+
+    @asset(deps=["a_key"])
+    def depends_on_source_asset():
+        return None
+
+    assert len(depends_on_source_asset.input_names) == 1
+    assert depends_on_source_asset.op.ins["a_key"].dagster_type.kind == DagsterTypeKind.NOTHING
+
+    res = materialize([depends_on_source_asset], resources={"io_manager": TestingIOManager()})
+    assert res.success
+
+
+def test_source_asset_deps_via_key():
+    a_source_asset = SourceAsset("a_key")  # noqa: F841
+
+    @asset(deps=[AssetKey("a_key")])
+    def depends_on_source_asset():
+        return None
+
+    assert len(depends_on_source_asset.input_names) == 1
+    assert depends_on_source_asset.op.ins["a_key"].dagster_type.kind == DagsterTypeKind.NOTHING
+
+    res = materialize([depends_on_source_asset], resources={"io_manager": TestingIOManager()})
+    assert res.success
+
+
+def test_interop():
+    @asset
+    def no_value_asset():
+        return None
+
+    @asset(io_manager_key="fs_io_manager")
+    def value_asset() -> int:
+        return 1
+
+    @asset(
+        deps=[no_value_asset],
+    )
+    def interop_asset(value_asset: int):
+        assert value_asset == 1
+
+    assert len(interop_asset.input_names) == 2
+    assert interop_asset.op.ins["no_value_asset"].dagster_type.kind == DagsterTypeKind.NOTHING
+    assert interop_asset.op.ins["value_asset"].dagster_type.kind == DagsterTypeKind.SCALAR
+
+    res = materialize(
+        [no_value_asset, value_asset, interop_asset],
+        resources={"io_manager": TestingIOManager(), "fs_io_manager": FilesystemIOManager()},
+    )
+    assert res.success
+
+
+def test_non_existent_asset_key():
+    @asset(deps=["not_real"])
+    def my_asset():
+        return None
+
+    res = materialize([my_asset], resources={"io_manager": TestingIOManager()})
+
+    assert res.success
+
+
```

Comment:
how do we handle when `deps` is set in addition to passing the same asset as an input to the function? reasonable error? I could see us potentially relaxing the error since the fn input trumps `deps` pretty clearly

Reply:
good call out - i will add a test

Reply:
you get the error 
```
E dagster._core.errors.DagsterInvalidDefinitionError: @op 'depends_on_upstream_asset' decorated function has parameter 'the_upstream_asset' that is one of the input_defs of type 'Nothing' which should not be included since no data will be passed for it.
```

Reply:
thats not great imo - we should at least have a better error and arguably permit it

Reply:
yeah this is the existing behavior though, so maybe we split it out as a separate PR?